src/cargo/ops/common_for_install_and_uninstall.rs

```diff
@@ -1,45 +1,481 @@
-use std::collections::{BTreeMap, BTreeSet};
+use std::collections::{btree_map, BTreeMap, BTreeSet};
 use std::env;
 use std::io::prelude::*;
 use std::io::SeekFrom;
 use std::path::{Path, PathBuf};
 
+use failure::{bail, format_err};
 use semver::VersionReq;
 use serde::{Deserialize, Serialize};
 
-use crate::core::PackageId;
-use crate::core::{Dependency, Package, Source, SourceId};
+use crate::core::{Dependency, Package, PackageId, Source, SourceId};
+use crate::ops::{self, CompileFilter, CompileOptions};
 use crate::sources::PathSource;
 use crate::util::errors::{CargoResult, CargoResultExt};
-use crate::util::{internal, Config, ToSemver};
-use crate::util::{FileLock, Filesystem};
-
-#[derive(Deserialize, Serialize)]
-#[serde(untagged)]
-pub enum CrateListing {
-    V1(CrateListingV1),
-    Empty(Empty),
+use crate::util::{Config, ToSemver};
+use crate::util::{FileLock, Filesystem, Freshness};
+
+/// On-disk tracking for which package installed which binary.
+///
+/// v1 is an older style, v2 is a new (experimental) style that tracks more
+/// information. The new style is only enabled with the `-Z install-upgrade`
+/// flag (which sets the `unstable_upgrade` flag). v1 is still considered the
+/// source of truth. When v2 is used, it will sync with any changes with v1,
+/// and will continue to update v1.
+///
+/// This maintains a filesystem lock, preventing other instances of Cargo from
+/// modifying at the same time. Drop the value to unlock.
+pub struct InstallTracker {
+    v1: CrateListingV1,
+    v2: CrateListingV2,
+    v1_lock: FileLock,
+    v2_lock: Option<FileLock>,
+    unstable_upgrade: bool,
```

Comment:
Could this perhaps be removed in favor of changing `v2` to an `Option`?

Reply:
I would prefer to keep it as-is, to make it easier to remove it later. If it was an option, there would need to be a bunch of unwrapping, and that would be a little more difficult to see exactly where the unstable behavior is located.

Reply:
Ok yeah that sounds reasonable to me!